src/main/java/org/broadinstitute/hellbender/utils/smithwaterman/SmithWatermanJavaAligner.java

```diff
@@ -64,37 +70,148 @@ public SmithWatermanAlignment align(final byte[] reference, final byte[] alterna
         Utils.nonNull(parameters);
         Utils.nonNull(overhangStrategy);
 
-        // avoid running full Smith-Waterman if there is an exact match of alternate in reference
-        int matchIndex = -1;
-        if (overhangStrategy == SWOverhangStrategy.SOFTCLIP || overhangStrategy == SWOverhangStrategy.IGNORE) {
-            // Use a substring search to find an exact match of the alternate in the reference
-            // NOTE: This approach only works for SOFTCLIP and IGNORE overhang strategies
-            matchIndex = Utils.lastIndexOf(reference, alternate);
-        }
-
         final SmithWatermanAlignment alignmentResult;
 
-        if (matchIndex != -1) {
-            // generate the alignment result when the substring search was successful
-            final List<CigarElement> lce = new ArrayList<>(alternate.length);
-            lce.add(makeElement(State.MATCH, alternate.length));
-            alignmentResult = new SWPairwiseAlignmentResult(AlignmentUtils.consolidateCigar(new Cigar(lce)), matchIndex);
-        }
-        else {
-            // run full Smith-Waterman
-            final int n = reference.length+1;
-            final int m = alternate.length+1;
-            final int[][] sw = new int[n][m];
-            final int[][] btrack=new int[n][m];
-
-            calculateMatrix(reference, alternate, sw, btrack, overhangStrategy, parameters);
-            alignmentResult = calculateCigar(sw, btrack, overhangStrategy); // length of the segment (continuous matches, insertions or deletions)
+        if (overhangStrategy == SWOverhangStrategy.SOFTCLIP || overhangStrategy == SWOverhangStrategy.IGNORE){
```

Comment:
I think this method has grown big and confusing, wherever possible in side of this if statement you should try to pull out the bodies of if statements into their own distinct methods that perhaps return an optional type. Like `findExactMatchBetweenStrings()`, then `findSingleIndelAlignments()` and so-fourth.